src/map/ai/helpers/pathfind.cpp

```diff
@@ -578,15 +580,45 @@ float CPathFind::GetRealSpeed()
         {
             realSpeed = 20;
         }
-        else if (m_POwner->animation == ANIMATION_ATTACK)
+
+        if (settings::get<bool>("map.USE_MOB_SPEED_BOOST_MULTIPLIER"))
         {
-            auto speedMod = settings::get<int8>("map.MOB_SPEED_MOD");
-            if (speedMod < -90)
+            if (auto* mobEntity = dynamic_cast<CMobEntity*>(m_POwner))
             {
-                speedMod = -90;
+                bool mobIsPlayerPet = mobEntity->PMaster && mobEntity->PMaster->objtype == TYPE_PC;
+
+                auto* battleTarget         = mobEntity->GetBattleTarget();
+                bool  targetWouldBeDrawnin = battleTarget && (distance(mobEntity->loc.p, battleTarget->loc.p) - battleTarget->m_ModelRadius) > mobEntity->GetMeleeRange();
+                bool  isEngaged            = mobEntity->animation == ANIMATION_ATTACK;
+
+                if (isEngaged && !mobIsPlayerPet && targetWouldBeDrawnin)
+                {
+                    // use default multiplier for most mobs
+                    int16 intMultiplier = settings::get<int16>("map.DEFAULT_MOB_SPEED_BOOST_MULTIPLIER");
+
+                    // unless the mob has a custom multiplier
+                    if (mobEntity->getMobMod(MOBMOD_SPEED_BOOST_MULT) > 0)
+                    {
+                        intMultiplier = mobEntity->getMobMod(MOBMOD_SPEED_BOOST_MULT);
+                    }
+
+                    // convert the integer based multiplier (based on 100) to float multiplier
+                    float floatMultiplier = intMultiplier / 100.0f;
+
+                    // if some weight penalty (like gravity) then cut the boost multiplier according to retail
+                    // (for mobs with default retail boost of 2.5 then boost becomes 1.20)
+                    if (mobEntity->getMod(Mod::MOVE_SPEED_WEIGHT_PENALTY) > 0)
+                    {
+                        floatMultiplier *= 0.48f;
+                    }
+
+                    // clamp the multiplier between 1x and 5x to prevent underflows and overflows
+                    floatMultiplier = std::clamp<float>(floatMultiplier, 1.0f, 5.0f);
+
+                    // apply the multiplier
+                    realSpeed *= floatMultiplier;
```

Comment:
Instead of clamping the multiplier, I would use a new int16 variable to clamp the ourput speed, to really ensure that custom speeds or weird future edge-cases dont overflow.
```c++
int16 newSpeed = std::clamp<int16>(realSpeed * floatMultiplier, 0, 255);

realSpeed = static_cast<uint8>(newSpeed)
```

Reply:
I'm sure theres a better way to do this though